pkg/pushsync/pushsync.go

```diff
@@ -3,3 +3,218 @@
 // license that can be found in the LICENSE file.
 
 package pushsync
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"time"
+
+	"github.com/ethersphere/bee/pkg/logging"
+	"github.com/ethersphere/bee/pkg/p2p"
+	"github.com/ethersphere/bee/pkg/p2p/protobuf"
+	"github.com/ethersphere/bee/pkg/pushsync/pb"
+	"github.com/ethersphere/bee/pkg/storage"
+	"github.com/ethersphere/bee/pkg/swarm"
+	"github.com/ethersphere/bee/pkg/topology"
+)
+
+const (
+	protocolName    = "pushsync"
+	protocolVersion = "1.0.0"
+	streamName      = "pushsync"
+)
+
+type PushSync struct {
+	streamer      p2p.Streamer
+	storer        storage.Storer
+	peerSuggester topology.ClosestPeerer
+	quit          chan struct{}
+
+	logger  logging.Logger
+	metrics metrics
+}
+
+type Options struct {
+	Streamer      p2p.Streamer
+	Storer        storage.Storer
+	ClosestPeerer topology.ClosestPeerer
+
+	Logger logging.Logger
+}
+
+var retryInterval = 10 * time.Second // time interval between retries
+
+func New(o Options) *PushSync {
+	ps := &PushSync{
+		streamer:      o.Streamer,
+		storer:        o.Storer,
+		peerSuggester: o.ClosestPeerer,
+		logger:        o.Logger,
+		metrics:       newMetrics(),
+		quit:          make(chan struct{}),
+	}
+
+	ctx := context.Background()
+	go ps.chunksWorker(ctx)
+
+	return ps
+}
+
+func (s *PushSync) Protocol() p2p.ProtocolSpec {
+	return p2p.ProtocolSpec{
+		Name:    protocolName,
+		Version: protocolVersion,
+		StreamSpecs: []p2p.StreamSpec{
+			{
+				Name:    streamName,
+				Handler: s.handler,
+			},
+		},
+	}
+}
+
+func (ps *PushSync) Close() error {
+	close(ps.quit)
+	return nil
+}
+
+func (ps *PushSync) handler(ctx context.Context, p p2p.Peer, stream p2p.Stream) error {
+	// handle chunk delivery from other node
+	_, r := protobuf.NewWriterAndReader(stream)
+	defer stream.Close()
+
+	var ch pb.Delivery
+
+	if err := r.ReadMsg(&ch); err != nil {
+		if err == io.EOF {
+			return nil
+		}
+		return err
+	}
+
+	// create chunk and store it in the local store
+	addr := swarm.NewAddress(ch.Address)
+	chunk := swarm.NewChunk(addr, ch.Data)
+	_, err := ps.storer.Put(ctx, storage.ModePutSync, chunk)
+	if err != nil {
+		return err
+	}
+
+	// push this to the closest node too
+	peer, err := ps.peerSuggester.ClosestPeer(addr)
+	if err != nil {
+		if errors.Is(err, topology.ErrWantSelf) {
+			// i'm the closest - nothing to do
+			return nil
+		}
+		return err
+	}
+	if err := ps.sendChunkMsg(ctx, peer, chunk); err != nil {
+		ps.metrics.SendChunkErrorCounter.Inc()
+		return err
+	}
+
+	return ps.storer.Set(ctx, storage.ModeSetSyncPush, chunk.Address())
+}
+
+func (ps *PushSync) chunksWorker(ctx context.Context) {
+	var chunks <-chan swarm.Chunk
+	var unsubscribe func()
+
+	// timer, initially set to 0 to fall through select case on timer.C for initialisation
+	timer := time.NewTimer(0)
+	defer timer.Stop()
+	chunksInBatch := -1
+
+	for {
+		select {
+		// handle incoming chunks
+		case ch, more := <-chunks:
```

Comment:
Sorry if I am asking the stupid question, I am not very familiar with the algorithm here.  When is `chunks` channel being written to? 🙈